crates/core/src/kernel/snapshot/visitors.rs

```diff
@@ -0,0 +1,192 @@
+//! Log replay visitors.
+//!
+//! Log replay visitors allow to extract additional actions during log replay.
+
+use std::collections::HashMap;
+
+use arrow::compute::{filter_record_batch, is_not_null};
+use arrow_array::{Array, Int64Array, RecordBatch, StringArray, StructArray};
+
+use super::ActionType;
+use crate::errors::DeltaResult;
+use crate::kernel::arrow::extract as ex;
+use crate::kernel::Transaction;
+
+/// Allows hooking into the reading of commit files and checkpoints whenever a table is loaded or updated.
+pub trait ReplayVisitor: std::fmt::Debug + Send + Sync {
+    fn as_any(&self) -> &dyn std::any::Any;
+
+    /// Process a batch
+    fn visit_batch(&mut self, batch: &RecordBatch) -> DeltaResult<()>;
+
+    /// return all relevant actions for the visitor
+    fn required_actions(&self) -> Vec<ActionType>;
+}
+
+/// Get the relevant visitor for the given action type
+pub fn get_visitor(action: &ActionType) -> Option<Box<dyn ReplayVisitor>> {
+    match action {
+        ActionType::Txn => Some(Box::new(AppTransactionVisitor::new())),
+        _ => None,
+    }
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct AppTransactionVisitor {
+    pub(crate) app_transaction_version: HashMap<String, Transaction>,
+}
+
+impl AppTransactionVisitor {
+    pub(crate) fn new() -> Self {
+        Self {
+            app_transaction_version: HashMap::new(),
+        }
+    }
+}
+
+impl AppTransactionVisitor {
+    pub fn merge(&self, map: &HashMap<String, Transaction>) -> HashMap<String, Transaction> {
+        let mut clone = map.clone();
+        for (key, value) in &self.app_transaction_version {
+            clone.insert(key.clone(), value.clone());
+        }
+        return clone;
+    }
+}
+
+impl ReplayVisitor for AppTransactionVisitor {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn visit_batch(&mut self, batch: &arrow_array::RecordBatch) -> DeltaResult<()> {
+        if batch.column_by_name("txn").is_none() {
+            return Ok(());
+        }
+
+        let txn_col = ex::extract_and_cast::<StructArray>(batch, "txn")?;
+        let filtered = filter_record_batch(batch, &is_not_null(txn_col)?)?;
+        let arr = ex::extract_and_cast::<StructArray>(&filtered, "txn")?;
+
+        let id = ex::extract_and_cast::<StringArray>(arr, "appId")?;
+        let version = ex::extract_and_cast::<Int64Array>(arr, "version")?;
+        let last_updated = ex::extract_and_cast_opt::<Int64Array>(arr, "lastUpdated");
+
+        for idx in 0..id.len() {
+            if id.is_valid(idx) {
+                let app_id = ex::read_str(id, idx)?;
+                if self.app_transaction_version.contains_key(app_id) {
+                    continue;
+                }
+                self.app_transaction_version.insert(
+                    app_id.to_owned(),
+                    Transaction {
+                        app_id: app_id.into(),
+                        version: ex::read_primitive(version, idx)?,
```

Comment:
Sorry for a drive-by comment, but: Does this mean this will only keeps the first version? since we continue if the hash map has the key

Reply:
Sorry, `test_app_txn_visitor` answered this in the affirmative.

Is that the correct behavior, though? The Delta protocol [says](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#transaction-identifiers)

> Delta only ensures that the **latest** version for a given appId is available in the table snapshot.

Reply:
The replay visits commits from latest to oldest. By keeping the first one we encounter, we are in fact only showing the latest.

Reply:
Oh I see! Thanks!